sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/objects.scala

```diff
@@ -96,3 +99,40 @@ object CombineTypedFilters extends Rule[LogicalPlan] {
     }
   }
 }
+
+/**
+ * Removes MapObjects when the following conditions are satisfied
+ *   1. Mapobject(e) where e is lambdavariable
+ *   2. the function will convert an expression MapObjects(e) to AssertNotNull(e)
```

Comment:
If we have `AssertNotNull`, then it's not a no-op and we can not eliminate this `MapObjects`. We may need to update `NullPropagation` to eliminate no-op `AssertNotNull`